### PR TITLE
Use weaver for semantic convention codegen

### PR DESCRIPTION
**Pull request by jsuereth**
I wasn't sure what you might want to change/fix about codegen when moving to weaver, so this attempts to keep the status quo as best as possible.

## TODOs

- [x] Import markdown / notes -> javadoc filter to the Jinja weaver uses.
- [x] Finish supporting *all* types of template keys or throw an exception if an unsupported one shows up.
- [x] Remove previous jinja template
- [x] filter unwanted semconv groups in JQ expression in weaver config.
- [x] figure out what to do w/ K8S convention camel case.
- [x] Use type_mapping in weaver.yaml instead of macros.
- [x] Add `stability: deprecated` attributes back to incubating classes.

## Possible Improvements thanks to weaver

- We could generate *all* the code (for incubating + stable) in the same weaver config, with the same shared set of templates.  I kept it as-is for now so you can more easily compare what's happening.
- We can register name conversions for method/class/etc. for easier to understand case conversions.
- We can register acronymns, for consistent capitalization.

### CODE REVIEW

File: buildscripts/templates/registry/incubating_java/IncubatingSemanticAttributes.java.j2
Comment (lquerel) [inline buildscripts/templates/registry/incubating_java/IncubatingSemanticAttributes.java.j2:43]: We could generalize the `type_mapping` section in weaver.yaml and support multiple mapping filters.
See https://github.com/open-telemetry/weaver/issues/156

Comment (jsuereth) [reply buildscripts/templates/registry/incubating_java/IncubatingSemanticAttributes.java.j2:43]: I like this idea.

File: buildscripts/templates/registry/java/SemanticAttributes.java.j2
Comment (jsuereth) [reply buildscripts/templates/registry/java/SemanticAttributes.java.j2:22]: I'll add a TODO for that

File: buildscripts/templates/registry/incubating_java/IncubatingSemanticAttributes.java.j2
Comment (jsuereth) [reply buildscripts/templates/registry/incubating_java/IncubatingSemanticAttributes.java.j2]: I was actually unable to, thanks to `enum` type.  PTAL.

File: buildscripts/templates/registry/incubating_java/IncubatingSemanticAttributes.java.j2
Comment (jsuereth) [reply buildscripts/templates/registry/incubating_java/IncubatingSemanticAttributes.java.j2:43]: I think we're still lacking some abiltiieshere.  Take a look at the diffs for notes.

File: buildscripts/templates/registry/java/SemanticAttributes.java.j2
Comment (jsuereth) [reply buildscripts/templates/registry/java/SemanticAttributes.java.j2]: Mostly fixed...

File: buildscripts/templates/registry/incubating_java/IncubatingSemanticAttributes.java.j2
Comment (jsuereth) [reply buildscripts/templates/registry/incubating_java/IncubatingSemanticAttributes.java.j2]: I used this.  It still doesn't handle enums properly.

File: semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/JvmIncubatingAttributes.java
Comment (jack-berg) [inline semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/JvmIncubatingAttributes.java:16]: Why are files like this missing missing?

File: semconv/src/main/java/io/opentelemetry/semconv/SchemaUrls.java
Comment (jack-berg) [inline semconv/src/main/java/io/opentelemetry/semconv/SchemaUrls.java:8]: This is not generated. We manually add constants to it with each release. There's a build check it make sure it has variables for each verion.

File: semconv/src/main/java/io/opentelemetry/semconv/SemanticAttributes.java
Comment (jack-berg) [inline semconv/src/main/java/io/opentelemetry/semconv/SemanticAttributes.java:26]: We're keeping a copy of this and `ResourceAttributes` around to ease the transition.

**Review by jack-berg — COMMENTED** [review]
What's the difference between build tools and weaver? Is the community trying to transition away from build tools?

File: semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/JvmIncubatingAttributes.java
Comment (jsuereth) [reply semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/JvmIncubatingAttributes.java:16]: This is now fixed

File: semconv/src/main/java/io/opentelemetry/semconv/SchemaUrls.java
Comment (jsuereth) [reply semconv/src/main/java/io/opentelemetry/semconv/SchemaUrls.java]: Fixed

File: semconv/src/main/java/io/opentelemetry/semconv/SemanticAttributes.java
Comment (jsuereth) [reply semconv/src/main/java/io/opentelemetry/semconv/SemanticAttributes.java]: Fixed

**Comment by jsuereth** [comment]
Update this, but we're now getting issues where weaver can't appropriately "javadoc-friendly" a comment, e.g.:

```
[ant:checkstyle] [ERROR] /home/joshuasuereth/src/open-telemetry/semantic-conventions-java/semconv/src/main/java/io/opentelemetry/semconv/ExceptionAttributes.java:26: <p> tag should be preceded with an empty line. [JavadocParagraph]
```

**Comment by jsuereth** [comment]
@jack-berg This is ready for review now.  We have the javadoc/comment fixes in place for codegen.

I haven't consolidated down to calling weaver ONCE in gradle yet, but I can do that in a follow up PR.

File: buildscripts/templates/registry/incubating_java/IncubatingSemanticAttributes.java.j2
Comment (jsuereth) [reply buildscripts/templates/registry/incubating_java/IncubatingSemanticAttributes.java.j2]: Fixed

File: buildscripts/templates/registry/java/weaver.yaml
Comment (jack-berg) [inline buildscripts/templates/registry/java/weaver.yaml:1]: Where can I read more about this file's contents?

Comment (jack-berg) [reply buildscripts/templates/registry/java/weaver.yaml:1]: https://github.com/open-telemetry/weaver/blob/main/docs/weaver-config.md

Comment (lquerel) [reply buildscripts/templates/registry/java/weaver.yaml:1]: And here https://github.com/open-telemetry/weaver/tree/main/crates/weaver_forge#comment-filter

File: build.gradle.kts
Comment (jack-berg) [inline build.gradle.kts:100]: Apparently the `otel/weaver` image doesn't publish for `arm64` platform yet. 
```suggestion
        "--rm",
        "--platform=linux/x86_64",
```

File: build.gradle.kts
Comment (jack-berg) [inline build.gradle.kts:115]: Meh.. we already have a `generateSemanticConventions` task which merged both into one. Also, it can be nice to have the ability to isolate to only generate stable or incubating.

File: buildscripts/checkstyle.xml
Comment (jack-berg) [inline buildscripts/checkstyle.xml:148]: What's the motivation here?

Comment (jack-berg) [reply buildscripts/checkstyle.xml:148]: Example of javadoc that fails the build without this. See the line with the lone `<p>` before the `<ul>` - weaver seems to use `<p> ` as a line break. 
```
public final class ErrorAttributes {
  /**
   * Describes a class of error the operation ended with.
   *
   * <p>Notes:
   *
   * <p>The {@code error.type} SHOULD be predictable, and SHOULD have low cardinality.
   *
   * <p>When {@code error.type} is set to a type (e.g., an exception type), its canonical class name
   * identifying the type within the artifact SHOULD be used.
   *
   * <p>Instrumentations SHOULD document the list of errors they report.
   *
   * <p>The cardinality of {@code error.type} within one instrumentation library SHOULD be low.
   * Telemetry consumers that aggregate data from multiple instrumentation libraries and
   * applications should be prepared for {@code error.type} to have high cardinality at query time
   * when no additional filters are applied.
   *
   * <p>If the operation has completed successfully, instrumentations SHOULD NOT set {@code
   * error.type}.
   *
   * <p>If a specific domain defines its own set of error identifiers (such as HTTP or gRPC status
   * codes), it's RECOMMENDED to:
   *
   * <p>
   *
   * <ul>
   *   <li>Use a domain-specific attribute
   *   <li>Set {@code error.type} to capture all errors, regardless of whether they are defined
   *       within the domain-specific set or not.
   * </ul>
   */
```

File: semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ContainerIncubatingAttributes.java
Comment (jack-berg) [inline semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ContainerIncubatingAttributes.java:114]: What's going on with the removal of these deprecated attributes?

File: build.gradle.kts
Comment (jsuereth) [reply build.gradle.kts:100]: I can fix that, can you open a feature request on weaver?

File: build.gradle.kts
Comment (jsuereth) [reply build.gradle.kts:115]: It just means you have to duplicate the weaver config yaml in both places for now

File: buildscripts/checkstyle.xml
Comment (jsuereth) [reply buildscripts/checkstyle.xml:148]: Yes - that's the motivation. I asked on the one CNCF slack channel about this, but didn't hear any feedback.

I'm unable to figure out how to resolve this succesfully, weaver is using <p> and the actual underlying data in the YAML has this empty line break.

File: build.gradle.kts
Comment (jack-berg) [reply build.gradle.kts:115]: Ah yeah would be good to DRY that up

File: buildscripts/checkstyle.xml
Comment (jack-berg) [reply buildscripts/checkstyle.xml:148]: > I'm unable to figure out how to resolve this succesfully, weaver is using

Looks like some of the message got truncated

FWIW, I don't think its a big deal to have empty paragraphs.

File: semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ContainerIncubatingAttributes.java
Comment (jsuereth) [reply semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ContainerIncubatingAttributes.java:114]: I'll take a look.  Likely I've filtered them out in my JQ filter for some reason...

File: build.gradle.kts
Comment (jack-berg) [reply build.gradle.kts:100]: https://github.com/open-telemetry/weaver/issues/345

File: semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ContainerIncubatingAttributes.java
Comment (jsuereth) [reply semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ContainerIncubatingAttributes.java:114]: It looks like all the experimental + deprecated attributes are not showing up here from the deprecated group in the registry, I'm investigating

File: buildscripts/templates/registry/incubating_java/weaver.yaml
Comment (jsuereth) [inline buildscripts/templates/registry/incubating_java/weaver.yaml:20]: @jack-berg - This is why we don't see deprecated *and* experimental things.  Setting to false, they'll all come back.

File: semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ContainerIncubatingAttributes.java
Comment (jsuereth) [reply semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ContainerIncubatingAttributes.java:114]: Ah right - If you recall the `client.id` vs. `client_id` debacle, the solution that was decided upon was to stop generating *deprecated* + *experimental* attributes in codegen.

Effectively, these are attributes that not only never stabilized, but also will never stabilize.

Including those types will cause a conflict in codegen between two generated code names.  We only guarantee uniqueness of generated code for the set of (stable attributes and non-deprecated experimental attributes).

So this is actually updating Java codegen here.  I can continue generating the deprecated experimental attributes, but two things:

(1) You'll need to carry any namespace-related conflict resolution that causes in the java codegen going forward.
(2) We likely want some way to denote "This is deprecated for the Stable thing" vs. "This is deprecated and never use it again".

cc @lmolkova to correct me if I confused anything on this resolution.

File: buildscripts/checkstyle.xml
Comment (jsuereth) [reply buildscripts/checkstyle.xml:148]: Great - I'd like to leave the changes as-is then.

File: buildscripts/templates/registry/incubating_java/IncubatingSemanticAttributes.java.j2
Comment (lmolkova) [inline buildscripts/templates/registry/incubating_java/IncubatingSemanticAttributes.java.j2:59]: members could be integers - check out `rpc.grpc.status_code` (and we have `print_member_value` helper to write `0` vs `"string"`)

File: semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ContainerIncubatingAttributes.java
Comment (lmolkova) [reply semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ContainerIncubatingAttributes.java:114]: How to resolve client.id vs client_id - https://github.com/open-telemetry/semantic-conventions/issues/1031#issuecomment-2173887975 - manually drop `messaging.client_id`.

Existing jinja and build-tools solution implements it already and here's the weaver-based approach in python
- [declare a list of excluded ones](https://github.com/open-telemetry/opentelemetry-python/blob/889f7df41ecca926c7847f3c0913b7b9bbc0dd62/scripts/semconv/templates/registry/weaver.yaml#L7)
- [exclude them in jinja](https://github.com/open-telemetry/opentelemetry-python/blob/889f7df41ecca926c7847f3c0913b7b9bbc0dd62/scripts/semconv/templates/registry/semantic_attributes.j2#L22) or in JQ (in python we decided to generate them as comments for transparency) 


WRT if we should generate deprecated+experimental attributes - we should. The motivation - dependency version conflicts - there is always a risk that semconv artifact version is resolved to something you don't expect. 

If experimental+deprecated attributes are not generated, it could mean that two parts of app that use schema_url vN and schema_url N+1 can become incompatible and can't coexist since there is no semconv artifact that contains both.

By generating experiemental+deprecated we make it ~~much easier~~ possible to update different parts of user app independently

Comment (jsuereth) [reply semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ContainerIncubatingAttributes.java:114]: Ok, this is now fixed.

File: buildscripts/templates/registry/incubating_java/IncubatingSemanticAttributes.java.j2
Comment (jsuereth) [reply buildscripts/templates/registry/incubating_java/IncubatingSemanticAttributes.java.j2:59]: It's funny - I fixed it in ONE template but not the other,thanks for catching this.

Comment (jsuereth) [reply buildscripts/templates/registry/incubating_java/IncubatingSemanticAttributes.java.j2]: Fixed

File: buildscripts/templates/registry/incubating_java/weaver.yaml
Comment (jsuereth) [reply buildscripts/templates/registry/incubating_java/weaver.yaml]: These should be back now.

**Review by lquerel — COMMENTED** [review]
Here are two minor suggestions:

1. Attributes are sorted by default, so most of the sort filters can be removed.
2. The file name definition can be simplified, as detailed below.

I recommend using the new, simpler method to specify file names. For example, in `IncubatingSemanticAttributes.java.j2`, instead of:

```jinja
{%- set my_file_name = ctx.root_namespace | pascal_case ~ "IncubatingAttributes.java" -%}
{{- template.set_file_name(my_file_name) -}}
```

you could define the file name directly in `weaver.yaml` within each template object like this:

```yaml
templates:
  - template: "incubating_java/IncubatingSemanticAttributes.java.j2"
    filter: ...
    application_mode: ...
    file_name: ctx.root_namespace | pascal_case ~ "IncubatingAttributes.java"
```

File: buildscripts/templates/registry/incubating_java/IncubatingSemanticAttributes.java.j2
Comment (jsuereth) [reply buildscripts/templates/registry/incubating_java/IncubatingSemanticAttributes.java.j2]: fixed

**Comment by jsuereth** [comment]
> Here are two minor suggestions:
> 
> 1. Attributes are sorted by default, so most of the sort filters can be removed.

Fixed

> 2. The file name definition can be simplified, as detailed below.
> 
> I recommend using the new, simpler method to specify file names. For example, in `IncubatingSemanticAttributes.java.j2`, instead of:
> 
> ```
> {%- set my_file_name = ctx.root_namespace | pascal_case ~ "IncubatingAttributes.java" -%}
> {{- template.set_file_name(my_file_name) -}}
> ```
> 
> you could define the file name directly in `weaver.yaml` within each template object like this:
> 
> ```yaml
> templates:
>   - template: "incubating_java/IncubatingSemanticAttributes.java.j2"
>     filter: ...
>     application_mode: ...
>     file_name: ctx.root_namespace | pascal_case ~ "IncubatingAttributes.java"
> ```

FYI - the `file_name` is actually a JINJA template, not an expression, so it's:

```
templates:
  - template: "incubating_java/IncubatingSemanticAttributes.java.j2"
    filter: ...
    application_mode: ...
    file_name: "{{ctx.root_namespace | pascal_case}}IncubatingAttributes.java"
```

File: semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/MessagingIncubatingAttributes.java
Comment (jack-berg) [inline semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/MessagingIncubatingAttributes.java:489]: I don't think that this is actually deprecated. Applies to several other attributes in this file as well. https://github.com/open-telemetry/semantic-conventions/blob/main/docs/messaging/azure-messaging.md#span-attributes

File: semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/HttpIncubatingAttributes.java
Comment (jack-berg) [inline semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/HttpIncubatingAttributes.java:317]: Something's fishy. This is still deprecated.

https://github.com/open-telemetry/semantic-conventions/blob/5eb37f706894718493af7a668971c651acb80812/docs/attributes-registry/http.md#deprecated-http-attributes

File: semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/HttpIncubatingAttributes.java
Comment (jack-berg) [inline semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/HttpIncubatingAttributes.java:250]: Why does this description go away? Is it gone from the YAML model? Or it something with the template?

Comment (jsuereth) [reply semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/HttpIncubatingAttributes.java:250]: It's how I `if-else` the template. I can add them back if you wanted, but thesee are attributes that have moved from Incubating -> Stable so I didn't keep the description.

Comment (jsuereth) [reply semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/HttpIncubatingAttributes.java:250]: fixed

File: semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/HttpIncubatingAttributes.java
Comment (jsuereth) [reply semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/HttpIncubatingAttributes.java:317]: I removed deprecated annotations from enum message definitions, adding them back in.

Comment (jsuereth) [reply semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/HttpIncubatingAttributes.java:317]: Fixed

File: semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ContainerIncubatingAttributes.java
Comment (lmolkova) [inline semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ContainerIncubatingAttributes.java:126]: oops

File: semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ErrorIncubatingAttributes.java
Comment (lmolkova) [inline semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ErrorIncubatingAttributes.java:57]: nit: this used to point to the enum type and now points to the attribute name.

Comment (jsuereth) [reply semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ErrorIncubatingAttributes.java]: Fixed

File: semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ContainerIncubatingAttributes.java
Comment (jsuereth) [reply semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ContainerIncubatingAttributes.java:126]: Fixed

**Review by lmolkova — APPROVED** [review]

File: semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/SystemIncubatingAttributes.java
Comment (jack-berg) [inline semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/SystemIncubatingAttributes.java:267]: These `@Deprecated` annotations should not go away

File: semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ArtifactIncubatingAttributes.java
Comment (jack-berg) [reply semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ArtifactIncubatingAttributes.java:78]: I don't think its a big deal. Your choice.

**Comment by jack-berg** [comment]
@jsuereth, @lmolkova, @lquerel, and @open-telemetry/java-approvers - I'll merge this tomorrow if there are no additional comments.

File: semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/CicdIncubatingAttributes.java
Comment (lquerel) [reply semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/CicdIncubatingAttributes.java:55]: The GH issue for tracking https://github.com/open-telemetry/weaver/issues/353

Comment (jsuereth) [reply semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/CicdIncubatingAttributes.java:55]: thanks @lquerel !

File: semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ArtifactIncubatingAttributes.java
Comment (jsuereth) [reply semconv-incubating/src/main/java/io/opentelemetry/semconv/incubating/ArtifactIncubatingAttributes.java:78]: Let's leave as-is for now.

Once we're on weaver (the important piece of this PR) there's incremental improvements we can begin making - but I'd like to focus on progress.

**Pull request merged**
